### ISSUE TITLE
spvTextToBinary to return a mapping of variables to IDs

### ISSUE THREAD

**Issue by umar456** [KhronosGroup/SPIRV-Tools#41]
Use Case:
Unit test in the validation rely on the error codes validate the result of the `spvValidate` function. Ideally the tests should check both the error code and the diagnostic message to verify that the test case is accurately passed. This is currently not possible because the named ID representation is lost during the Text to Binary conversion.

The `spvTextToBinary` function should return a map which contains a mapping from ID to ID names used in the textual representation of the SPIR-V.

**Comment by dmircevski** [comment]
This would be nice to have, though it's not essential for checking the diagnostic messages even now.  The message text around %IDs can be checked, and `OpName` instructions (where present) could be used to map %IDs to source names.

**Comment by umar456** [comment]
Good point. I should be able to create diagnostic messages based on the (OpName) commands and verify the output using this approach.

**Comment by dneto0** [comment]
Keep in mind that OpName instructions are optional.
Also, the validator is checking binaries coming from any front end, which could be much more than the assembler in SPIR-V Tools.
So I see using the OpName values in validator diagnostics as nice-to-have, but not essential.

**Comment by dmircevski** [comment]
Similarly, not every front end can furnish spvValidate() with a %ID mapping
proposed in OP.

On Thu, 3 Dec 2015 at 14:43 David Neto <notifications@github.com> wrote:

> Keep in mind that OpName instructions are optional.
> Also, the validator is checking binaries coming from any front end, which
> could be much more than the assembler in SPIR-V Tools.
> So I see using the OpName values in validator diagnostics as nice-to-have,
> but not essential.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KhronosGroup/SPIRV-Tools/issues/41#issuecomment-161759594>
> .
>
-- 

-- 
Sent from a mobile device.

**Comment by umar456** [comment]
I understand. I was plan on formatting the diagnostic message to display the id and only display the name if an OpName command references the ID. For example

No OpName
```
ID 32 has not been defined
```
With OpName
```
ID 32[foo] has not been defined
```

This should be pretty easy to implement. I can then use regex to parse the message to verify the unit tests.

**Comment by dneto0** [comment]
Yes, that would work.

Alternately, you could use the same strategy the assembler tests use: rely on the fact that the assembler assigns Id indices via a counter, starting at 1.

See, for example, some disassembler diagnostics tests at https://github.com/KhronosGroup/SPIRV-Tools/blob/master/test/BinaryToText.cpp#L181
Those tests first assemble some text, then disassemble text but fail with a message. (They use TextToBinaryTestBase::EncodeSuccessfullyDecodeFailed)

**Comment by umar456** [comment]
That sounds like it would be dependent on the implementation. I wouldn't feel comfortable writing tests like that. It could also become complicated once we test more complicated cases.

**Comment by dj2** [comment]
I'm not sure if there is anything to do here,`spirv-dis` does the name mapping when possible already. Can this be closed?
